core/src/main/scala/kafka/log/Log.scala

```diff
@@ -256,7 +256,8 @@ class Log(@volatile private var _dir: File,
           val topicPartition: TopicPartition,
           val producerStateManager: ProducerStateManager,
           logDirFailureChannel: LogDirFailureChannel,
-          private val hadCleanShutdown: Boolean = true) extends Logging with KafkaMetricsGroup {
+          private val hadCleanShutdown: Boolean = true,
+          val keepPartitionMetadataFile: Boolean = true) extends Logging with KafkaMetricsGroup {
```

Comment:
Could we add the new param to the javadoc? In the javadoc, it would be useful to explain a bit how this helps with re-upgrade.

Reply:
good point. will do!

Reply:
Something like this work for an explanation?
```
* boolean flag to indicate whether the partition metadata file should be kept in the 
* log directory. A partition.metadata file is only created when the controller's 
* inter-broker protocol version is at least 2.8. This file will persist the topic ID on
* the broker. If inter-broker protocol is downgraded below 2.8, a topic ID may be lost
* and a new ID generated upon re-upgrade. If the inter-broker protocol version is below
* 2.8, partition.metadata will be deleted to avoid ID conflicts upon re-upgrade. 
 ```

Reply:
Yes, looks good to me.